patches/server/0918-Add-damage-source-wrapper.patch

```diff
@@ -0,0 +1,826 @@
++    static {
++        badRespawnPointExplosion().bindStatic();
++    }
++    // Paper end
+     private boolean damageHelmet;
+     private boolean bypassArmor;
+     private boolean bypassInvul;
+     private boolean bypassMagic;
+     private boolean bypassEnchantments;
+-    private float exhaustion = 0.1F;
++    public float exhaustion = 0.1F; // Paper private -> public - Damage source wrapper
+     private boolean isFireSource;
+     private boolean isProjectile;
+     private boolean scalesWithDifficulty;
+@@ -64,6 +70,12 @@ public class DamageSource {
+         return this;
+     }
+     // CraftBukkit end
++    // Paper start - Damage source wrapper
++    public DamageSource bindStatic() {
++        staticDamageBinder.put(msgId, io.papermc.paper.entity.damage.DamageOriginImpl.fromNMS(this));
++        return this;
++    }
++    // Paper end
+     // Paper start - add critical damage API
+     private boolean critical;
+     public boolean isCritical() {
+@@ -202,34 +214,46 @@ public class DamageSource {
+         return null;
+     }
+ 
+-    protected DamageSource bypassArmor() {
++    // Paper start - Damage source wrapper
++    public DamageSource bypassArmor() {
++        return bypassArmor(false);
++    }
++
++    public DamageSource bypassArmor(boolean ignoreExhaustion) { // Paper protected -> public - Damage source wrapper
+         this.bypassArmor = true;
+-        this.exhaustion = 0.0F;
++        if(!ignoreExhaustion) this.exhaustion = 0.0F;
++        // Paper end
+         return this;
+     }
+ 
+-    protected DamageSource damageHelmet() {
++    public DamageSource damageHelmet() { // Paper protected -> public - Damage source wrapper
+         this.damageHelmet = true;
+         return this;
+     }
+ 
+-    protected DamageSource bypassInvul() {
++    public DamageSource bypassInvul() { // Paper protected -> public - Damage source wrapper
+         this.bypassInvul = true;
+         return this;
+     }
+ 
+-    protected DamageSource bypassMagic() {
++    // Paper start - Damage source wrapper
++    public DamageSource bypassMagic() {
++        return bypassMagic(false);
++    }
++
++    public DamageSource bypassMagic(boolean ignoreExhaustion) { // Paper protected -> public - Damage source wrapper
+         this.bypassMagic = true;
+-        this.exhaustion = 0.0F;
++        if(!ignoreExhaustion) this.exhaustion = 0.0F;
++        // Paper end
+         return this;
+     }
+ 
+-    protected DamageSource bypassEnchantments() {
++    public DamageSource bypassEnchantments() { // Paper protected -> public - Damage source wrapper
+         this.bypassEnchantments = true;
+         return this;
+     }
+ 
+-    protected DamageSource setIsFire() {
++    public DamageSource setIsFire() {// Paper protected -> public - Damage source wrapper
+         this.isFireSource = true;
+         return this;
+     }
```

Comment:
For the builder methods that modify exhaustion, simply store the previous exhaustion and make sure it's set back to what it was before.

capture exhaustion
bypassMagic()
set exhaustion to what was captured

Reply:
This isn't needed cos any builder will build a new damage origin which will called the bypass methods before defining the exhaustion amount